### PR TITLE
Create CONTRIBUTING file

### LINKED ISSUE #138: Create a contribution guide
## Problem
Since this repository will be treated as a proper open-source project, it should have an easy-to-follow contribution guide to enable other people contribute and improve the library and codebase.

## Solution suggestion
Create a contributing guide, use possible tools and get inspired by successful open-source projects

closes #101

### PR DESCRIPTION
**Pull request by fpindej**
closes #138 

## Summary

Maintaining an open-source project is not too difficult, but it has to be done properly and people need to know how to contribute. This PR introduces a `CONTRIBUTING.MD`, which is a very common practice for open-source repositories.

## Solution

- [x] generate a boilerplate file using  [contributing.md](https://contributing.md/) generator 
- [ ] validate and edit the file:
     - [x] Contributing to FuncSharp
     - [x] Table of Contents
     - [ ] I Have a Question
     - [ ] I Want to Contribute
     - [x] Legal Notice
     - [ ] Reporting Bugs
     - [x] Before Submitting a Bug Report
     - [x] How Do I Submit a Good Bug Report?
     - [ ] Suggesting enhancements
     - [ ] Before Submitting an Enhancement
     - [ ] How Do I Submit a Good Enhancement Suggestion?
     - [x] Your First Code Contribution
     - [x] Environment setup
     - [x] Choose an Issue
     - [x] Submit a Pull Request
     - [ ] Improving The Documentation
     - [ ] Styleguides
     - [x] Commit Messages

### CODE REVIEW

File: docs/CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+-->
+
+### Improving The Documentation
+<!-- TODO
+Updating, improving and correcting the documentation
+
+-->
+
+## Styleguides
+### Commit Messages
```
Comment (fpindej) [inline docs/CONTRIBUTING.md:148]: I've been using this style for a while and it works super well:
image

Comment (fpindej) [reply docs/CONTRIBUTING.md:148]: Especially when combined with a rebase/interactive rebase. The history becomes super clean 😊 Almost like a mind map or a guide 😄

File: docs/CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+- [Improving The Documentation](#improving-the-documentation)
+- [Styleguides](#styleguides)
+- [Commit Messages](#commit-messages)
+- [Join The Project Team](#join-the-project-team)
+
+
+
+## I Have a Question
+
+> If you want to ask a question, we assume that you have read the available [Documentation]().
```
Comment (KaliCZ) [inline docs/CONTRIBUTING.md:30]: Hah - maybe let's take the link out for now.

File: docs/CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+Before you ask a question, it is best to search for existing [Issues](https://github.com/MewsSystems/FuncSharp/issues) that might help you. In case you have found a suitable issue and still need clarification, you can write your question in this issue. It is also advisable to search the internet for answers first.
+
+If you then still feel the need to ask a question and need clarification, we recommend the following:
+
+- Open an [Issue](https://github.com/MewsSystems/FuncSharp/issues/new).
```
Comment (KaliCZ) [inline docs/CONTRIBUTING.md:36]: Speaking of - we should have an issue template.

File: docs/CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+- IRC
+- Slack
+- Gitter
+- Stack Overflow tag
+- Blog
+- FAQ
+- Roadmap
+- E-Mail List
+- Forum
+-->
```
Comment (KaliCZ) [inline docs/CONTRIBUTING.md:55]: Yeah I don't think we'd need this section in the near future.

File: docs/CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+
+### Reporting Bugs
+
+<!-- omit in toc -->
+#### Before Submitting a Bug Report
+
+A good bug report shouldn't leave others needing to chase you up for more information. Therefore, we ask you to investigate carefully, collect information and describe the issue in detail in your report. Please complete the following steps in advance to help us fix any potential bug as fast as possible.
+
+- Make sure that you are using the latest version.
+- Determine if your bug is really a bug and not an error on your side e.g. using incompatible environment components/versions (Make sure that you have read the [documentation](). If you are looking for support, you might want to check [this section](#i-have-a-question)).
```
Comment (KaliCZ) [inline docs/CONTRIBUTING.md:70]: Again the links :)

File: docs/CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+- OS, Platform and Version (Windows, Linux, macOS, x86, ARM)
+- Version of the interpreter, compiler, SDK, runtime environment, package manager, depending on what seems relevant.
+- Possibly your input and the output
+- Can you reliably reproduce the issue? And can you also reproduce it with older versions?
+
+<!-- omit in toc -->
+#### How Do I Submit a Good Bug Report?
+
+> You must never report security related issues, vulnerabilities or bugs including sensitive information to the issue tracker, or elsewhere in public. Instead sensitive bugs must be sent by email to <>.
+<!-- You may add a PGP key to allow the messages to be sent encrypted as well. -->
```
Comment (KaliCZ) [inline docs/CONTRIBUTING.md:84]: I'd simplify 

```suggestion
> Make sure not to include sensitive information when asking questions or submitting bugs anywhere in public.
```

File: docs/CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+- You may want to **include screenshots and animated GIFs** which help you demonstrate the steps or point out the part which the suggestion is related to. You can use [this tool](https://www.cockos.com/licecap/) to record GIFs on macOS and Windows, and [this tool](https://github.com/colinkeenan/silentcast) or [this tool](https://github.com/GNOME/byzanz) on Linux. <!-- this should only be included if the project has a GUI -->
+- **Explain why this enhancement would be useful** to most FuncSharp users. You may also want to point out the other projects that solved it better and which could serve as inspiration.
+
+<!-- You might want to create an issue template for enhancement suggestions that can be used as a guide and that defines the structure of the information to be included. If you do so, reference it here in the description. -->
+
+### Your First Code Contribution
+
+#### Environment setup
+
+All You need is a .NET 6 SDK, an IDE of your choice and you're good to go! 🎉
```
Comment (KaliCZ) [inline docs/CONTRIBUTING.md:131]: Actually, you need an IDE capable of executing T4 scripts. 
Unfortunately even rider itself is not capable of running the T4 scripts. You'd need to get the dlls from microsoft to make it work. I solved this by also installing visual studio and simply not using it.

File: docs/CONTRIBUTING.md [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+
+## Styleguides
+### Commit Messages
+
+We propose a recommended approach for composing and comprehending commits: 
+
+```
+"This commit will <YOUR_COMMIT_MESSAGE>"
+``` 
+serving as a clear template. For instance, a commit could be `"This commit will <Introduce new data type>"`, providing a concise insight into the purpose of the commit.
```
Comment (KaliCZ) [inline docs/CONTRIBUTING.md:155]: I wouldn't have the `This commit will` prefix - You even don't have it in the screenshot 😄

**Review by KaliCZ — APPROVED** [review]
I don't know if it's some sort of a convention to put it in a docs folder. But I'd have the contributing file next to the readme and license file - in the root folder.

Love it though. Good first version.

**Pull request closed without merging**